guides/preferred_patterns.md

```diff
@@ -0,0 +1,46 @@
+# Preferred Coding Patterns
+
+The presented patterns are meant as a general guide but should not be understood 
+as rules to follow blindly or use as a knockout argument or sledgehammer 
+approach to everything. Each problem deserves to be judged individually 
+considering all its aspects and complexity when choosing a fitting solution. 
+
+However, when in doubt which approach might be more fitting the guides provide
+a reasoning and direction with the goal of a more uniform codebase which can
+benefit from the advantages that emerge when using the described patterns.
+
+## Overview
+| Pattern | Prefer | Over                    |
+|---------|---------------------|-------------------------|
+| Consider collection unmodifiable | copy before modification | copy before returning |
+
+## Patterns
+
+Provides more detailed reasoning why and when to use a certain pattern. 
+
+### Consider Collections as _unmodifiable_
+When working with collections errors can arise from trying to manipulate an
+_unmodifiable_ collection. One way to defend against this is to make copies of
+collections to be sure they can be modified by a consumer. This often results
+in collections being copied multiple times over as they pass through the layers
+of abstractions and the application architecture. The other approach to this
+is to generally consider any collection that wasn't created within or as part 
+of the code that wants to modify the collection as _unmodifiable_. 
+
+In practice this means any collection received from a virtual method call or
```

Comment:
what do you mean by virtual here? Maybe remove the word

Reply:
I wsa thinking in JVM terms. There _invokevirtual_ means calling a potentially polymorphic method. I used this to emphasis that you can't really know what implementation you will get and therefore have no way of knowing if it is possible to modify the returned collection. But it is also misleading as we should also include results e.g. of a static method in a util that is not dynamically bound and still we should not assume a specific implementation. Either we remove the word or find a better way of saying what I described.

Reply:
```suggestion
In practice this means any collection received from a non-private method call or
```

Reply:
Thank you for the explanation 😄 I get it now 😅